File: AndroidXPatternsApp/jmvvm-app/src/main/java/com/woowrale/jmvvm/ui/search/local/LocalSearchPresenter.java
package com.woowrale.jmvvm.ui.search.local;

import android.util.Log;

import com.jakewharton.rxbinding2.widget.TextViewTextChangeEvent;
import com.woowrale.jmvvm.BuildConfig;
import com.woowrale.jmvvm.data.model.Contact;
import com.woowrale.jmvvm.data.repository.ApiService;
import com.woowrale.jmvvm.ui.base.BasePresenter;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.observers.DisposableObserver;
import io.reactivex.observers.DisposableSingleObserver;
import io.reactivex.schedulers.Schedulers;

@Singleton
public class LocalSearchPresenter extends BasePresenter<LocalSearchView> {

    private static final String TAG = LocalSearchPresenter.class.getSimpleName();

    private ApiService apiService;

    @Inject
    public LocalSearchPresenter(ApiService apiService) {
        this.apiService = apiService;
    }

    public DisposableObserver<TextViewTextChangeEvent> searchContacts() {
        return new DisposableObserver<TextViewTextChangeEvent>() {
            @Override
            public void onNext(TextViewTextChangeEvent textViewTextChangeEvent) {
                Log.d(TAG, "Search query: " + textViewTextChangeEvent.text());
                getView().getmAdapter().getFilter().filter(textViewTextChangeEvent.text());
            }

            @Override
            public void onError(Throwable e) {
                Log.e(TAG, "onError: " + e.getMessage());
            }

            @Override
            public void onComplete() {

            }
        };
    }

    public void fetchContacts(String source) {
        getView().getDisposable().add(apiService
                .getContacts(BuildConfig.GET_CONTACTS, source, null)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeWith(new DisposableSingleObserver<List<Contact>>() {
                    @Override
                    public void onSuccess(List<Contact> contacts) {
                        getView().getContactsList().clear();
                        getView().getContactsList().addAll(contacts);
                        getView().getmAdapter().notifyDataSetChanged();
                    }

                    @Override
                    public void onError(Throwable e) {

                    }
                }));
    }

}

File: AndroidPatternsApp/kmvp-app/src/main/java/com/woowrale/mvp/ui/search/local/LocalSearchPresenter.kt
package com.woowrale.mvp.ui.search.local

import android.util.Log
import com.jakewharton.rxbinding2.widget.TextViewTextChangeEvent
import com.woowrale.mvp.BuildConfig
import com.woowrale.mvp.data.model.Contact
import com.woowrale.mvp.data.repository.ApiService
import com.woowrale.mvp.ui.base.BasePresenter
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.observers.DisposableObserver
import io.reactivex.observers.DisposableSingleObserver
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class LocalSearchPresenter@Inject constructor() : BasePresenter<LocalSearchView>() {

    private val TAG = LocalSearchPresenter::class.java.simpleName

    fun searchContacts(): DisposableObserver<TextViewTextChangeEvent> {
        return object : DisposableObserver<TextViewTextChangeEvent>() {
            override fun onNext(textViewTextChangeEvent: TextViewTextChangeEvent) {
                Log.d(TAG, "Search query: " + textViewTextChangeEvent.text())
                getView().getmAdapter().filter.filter(textViewTextChangeEvent.text())
            }

            override fun onError(e: Throwable) {
                Log.e(TAG, "onError: " + e.message)
            }

            override fun onComplete() {

            }
        }
    }


    fun fetchContacts(source: String) {
        getView().getDisposable().add(ApiService.services()
            .getContacts(BuildConfig.GET_CONTACTS, source, "")
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeWith(object : DisposableSingleObserver<List<Contact>>() {
                override fun onSuccess(contacts: List<Contact>) {
                    getView().setContactsList(contacts)
                    getView().getmAdapter().notifyDataSetChanged()
                }

                override fun onError(e: Throwable) {

                }
            })
        )
    }
}